processing/src/main/java/org/apache/druid/frame/processor/AwaitAnyWidget.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.frame.processor;
+
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+import com.google.errorprone.annotations.concurrent.GuardedBy;
+import it.unimi.dsi.fastutil.ints.IntIterator;
+import it.unimi.dsi.fastutil.ints.IntSet;
+import org.apache.druid.frame.channel.ReadableFrameChannel;
+import org.apache.druid.java.util.common.concurrent.Execs;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Helper used by {@link FrameProcessorExecutor#runFully} when workers return {@link ReturnOrAwait#awaitAny}.
+ *
+ * The main idea is to reuse listeners from previous calls to {@link #awaitAny(IntSet)} in cases where a particular
+ * channel has not receieved any input since the last call. (Otherwise, listeners would pile up.)
+ */
+public class AwaitAnyWidget
+{
+  private final List<ReadableFrameChannel> channels;
+
+  @GuardedBy("listeners")
+  private final List<Listener> listeners;
+
+  public AwaitAnyWidget(final List<ReadableFrameChannel> channels)
+  {
+    this.channels = channels;
+    this.listeners = new ArrayList<>(channels.size());
+
+    for (int i = 0; i < channels.size(); i++) {
+      listeners.add(null);
+    }
+  }
+
+  public ListenableFuture<?> awaitAny(final IntSet awaitSet)
```

Comment:
nit: awaitChannelNumberSet ?

Reply:
Also, can you add documentation about what does a future with a null value mean ?

Reply:
I added some comments to hopefully clear this up. I did keep `awaitSet` though, since it echoes a similar concept `ReturnOrAwait`.